chore: Change the confusing parameter `position` to a clear name. #637

JSpiner opened the pull request:
Thank you for opening a Pull Request!

---

Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open a GitHub issue as a bug/feature request before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #637 🦕

It may be confusing that it is automatically updated by the `position` parameter. I renamed it to `initialPosition` to make it clear that it only works with the initial value.
(The compose foundation libraries also add the init prefix to the initial values ​​that go into rememberXXX.)


----

MarkerState has rememberMarkerState for use in compose.

```kotlin
@Composable
public fun rememberMarkerState(
    key: String? = null,
    position: LatLng = LatLng(0.0, 0.0)
): MarkerState = rememberSaveable(key = key, saver = MarkerState.Saver) {
    MarkerState(position)
}
```

```kotlin
// wrong usage example
@Composable
public fun MyGoogleMapScreen(stationPosition: LatLng) {
    val myMarkerState = rembmerMarkerState(position = stationPosition)
}
```

But looking at the example above, it's easy to get confused.
This is because there is a risk of misunderstanding that the value entered as a parameter to remember (position in this case) acts as a key that is automatically reflected when the value changes.
I also had a hard time because the position wasn't updated.
To avoid this misunderstanding, compose foundation libraries add an inital prefix to rememberXXX functions.

```kotlin
// compose library example
// rememberScrollState from `androidx.compose.foundation`
@Composable
fun rememberScrollState(initial: Int = 0): ScrollState {
    return rememberSaveable(saver = ScrollState.Saver) {
        ScrollState(initial = initial)
    }
}
```

```kotlin
// rememberLazyListState from `androidx.compose.foundation.lazy`
@Composable
fun rememberLazyListState(
    initialFirstVisibleItemIndex: Int = 0,
    initialFirstVisibleItemScrollOffset: Int = 0
): LazyListState {
    return rememberSaveable(saver = LazyListState.Saver) {
        LazyListState(
            initialFirstVisibleItemIndex,
            initialFirstVisibleItemScrollOffset
        )
    }
}
```

```kotlin
// rememberLazyGridState from `androidx.compose.foundation.lazy.grid`
@Composable
fun rememberLazyGridState(
    initialFirstVisibleItemIndex: Int = 0,
    initialFirstVisibleItemScrollOffset: Int = 0
): LazyGridState {
    return rememberSaveable(saver = LazyGridState.Saver) {
        LazyGridState(
            initialFirstVisibleItemIndex,
            initialFirstVisibleItemScrollOffset
        )
    }
}
```

So my suggestion is to change the name from position to initialPosition to reduce confusion.

JSpiner commented:
I resolved the conflict.

bubenheimer commented on maps-compose/src/main/java/com/google/maps/android/compose/Marker.kt:
```diff
@@ -155,12 +155,12 @@ public class MarkerState private constructor(position: LatLng) {
         public operator fun invoke(
-            position: LatLng = LatLng(0.0, 0.0)
-        ): MarkerState = MarkerState(position)
+            initialPosition: LatLng = LatLng(0.0, 0.0)
```
I don't see the rationale for changing the parameter name here. Makes sense for rememberMarkerState, but not here. If you end up recomposing MarkerState.invoke(position) you will get a new MarkerState object.

JSpiner replied:
You are right. This part didn't need to be updated. 
However, I completely reverted and rewrote this part based on @kikoso 's opinion. 
Please review again. 🙇

JSpiner commented:
@kikoso @bubenheimer 

Following @kikoso 's suggestion, I kept the existing `rememberMarkerState` function without breaking changes.
All existing changes have been reverted and rewritten.

- Instead of leaving the existing 'rememberMarkerState' function as is, I added a comment that the state is not updated depending on the parameter.
- Added 'rememberUpdatedMarkerState' which updates the state based on the parameter.
- I wrote a test code for 'rememberUpdatedMarkerState'.

I think it would be a good idea to deprecate rememberMarkerState like below. I'd like to hear your opinions.

```kotlin
@Composable
@Deprecated(
    message = "Use 'rememberUpdatedMarkerState' instead - It may be confusing to think " +
            "that the state is automatically updated as the position changes, " +
            "so it will be changed.",
    replaceWith = ReplaceWith(
        expression = """
            val markerState = rememberUpdatedMarkerState()
        """
    )
)
public fun rememberMarkerState(
    key: String? = null,
    position: LatLng = LatLng(0.0, 0.0)
): MarkerState = rememberSaveable(key = key, saver = MarkerState.Saver) {
    MarkerState(position)
}
```

bubenheimer commented:
@JSpiner I agree with deprecating rememberMarkerState, it's a misleading API. The `@ReplaceWith` part seems off, though; I think it should give the original implementation of rememberMarkerState, to preserve semantics, and because users need to know where to find the `Saver`.

I _disagree_ with the changed implementation of rememberUpdatedMarkerState, basing it on rememberSaveable instead of remember. This is invalid for the examples where it is used, and does not align with the rationale from the posts in https://github.com/googlemaps/android-maps-compose/issues/637#issuecomment-2415820993

JSpiner commented:
@bubenheimer 

Thanks for comments.

> I disagree with the changed implementation of rememberUpdatedMarkerState, basing it on rememberSaveable instead of remember. 

Sorry, my mistake. I changed it to use the original implementation.
[356d429](https://github.com/googlemaps/android-maps-compose/pull/638/commits/356d429b8b194ac22a1c8fde15b50cdd47dd973a)

> I think it should give the original implementation of rememberMarkerState

I apply it. thanks
17f38572dc5ed425885e9492c7f5064d0ce78904

bubenheimer commented on app/src/androidTest/java/com/google/maps/android/compose/GoogleMapViewTests.kt:
```diff
@@ -281,6 +282,30 @@ class GoogleMapViewTests {
             .performClick()
     }
 
+    @Test
```
Does this test actually pass?

JSpiner replied:
yes it is.

When I run the test with the existing `rememberMarkerState`, the test fails as below.
<img width="825" alt="스크린샷 2024-12-16 23 14 02" src="https://github.com/user-attachments/assets/af46bf75-4a30-443c-836b-983ab60425f1" />

When I run the test with `rememberUpdatedMarkerState`, the test succeeds as shown below.
<img width="817" alt="스크린샷 2024-12-16 23 15 52" src="https://github.com/user-attachments/assets/6c8fe273-771c-4b95-8575-95e1a2eb0bbd" />

I can't verify this because there is no CI in this repo, but I think you can verify it by running the tests locally.

bubenheimer commented on maps-compose/src/main/java/com/google/maps/android/compose/Marker.kt:
```diff
@@ -182,15 +183,43 @@ public class MarkerState private constructor(position: LatLng) {
+ * This function updates the state value according to the update of the input parameter,
+ * like 'rememberUpdatedState'.
+ *
+ * This cannot be used to preserve info window visibility across configuration changes.
```
```suggestion
 * This cannot be used to preserve state across configuration changes.
```

bubenheimer reviewed (changes requested):
A maintainer should review this as well.

JSpiner commented:
> Also, @JSpiner : did you have the chance to take a look at the [CLA agreement](https://cla.developers.google.com/)?

@kikoso I agreed to the CLA agreement. When I looked at the commit logs, my secondary git account name was included and was processed as a non-agreement. I re-organized them by rebasing. Thanks.

bubenheimer commented:
> When merging @bubenheimer 's suggestion on github, he's name was included in the commit author [5761aaf](https://github.com/googlemaps/android-maps-compose/commit/5761aafbc991ea4739a48273bc28da6d60010f7f). Would you also agree to the CLA?

Done

JSpiner commented:
@kikoso 
Hi, CLA validation passed. Can this be merged?
